packages/use-sync-external-store/src/useSyncExternalStore.js

```
@@ -7,6 +7,182 @@
+  useLayoutEffect(() => {
+    const handleStoreChange = () => {
+      // TODO: Because there is no cross-renderer API for batching updates, it's
+      // up to the consumer of this library to wrap their subscription event
+      // with unstable_batchedUpdates. Should we try to detect when this isn't
+      // the case and print a warning in development?
+
+      // Track the current state of the store. We track this within the
+      // subscription handler instead of an effect to account for updates that
+      // happen in between render and commit -- a similar scenario to the one
+      // described above regarding getSnapshot. So this represents the most
+      // recent store state, not the most recent React state.
+
+      // If getSnapshot was changed during render and we haven't reached the
+      // commit effect yet, then it will be null. Otherwise, we can use it to
+      // bail out early.
+      const latestGetSnapshot = inst.getSnapshot;
+      if (latestGetSnapshot !== null) {
+        const prevValue = inst.value;
+        try {
+          const nextValue = getSnapshot();
+          if (!is(prevValue, nextValue)) {
+            // Force a re-render.
+            inst.value = nextValue;
+            setVersion(bumpVersion);
+          }
+        } catch (error) {
+          // Force a re-render.
+          setVersion(bumpVersion);
+        }
+      } else {
+        // Force a re-render.
+        setVersion(bumpVersion);
+      }
+    };
+
+    // When we initially subscribe, we need to check if there was a mutation in
+    // between render and commit. If there was, we may need to re-render.
+    // Subsequent changes will be detected in the subscription handler.
+    const valueDuringRender = value;
+    try {
+      const nextValue = getSnapshot();
+      if (!is(valueDuringRender, nextValue)) {
+        // Force a re-render.
+        inst.value = nextValue;
+        setVersion(bumpVersion);
+      }
+    } catch (error) {
```

What errors do we expect here? Is it proper error or also errors that happen due to tearing that can be fixed? Should this intentionally silence errors or no?

---

Since we also schedule an update, it will throw again during render so you can catch it with an error boundary. Same thing that we do for reducers.